group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/GeneralUniformAssignmentBuilder.java

```diff
@@ -14,17 +14,828 @@
+     *
+     * Only populated with partitions that weren't retained due to a rack mismatch when rack aware strategy is used.
+     */
+    private final Map<TopicIdPartition, String> currentPartitionOwners;
+
+    /**
+     * Tracks the owner of each partition in the target assignment.
+     */
+    private final Map<TopicIdPartition, String> partitionOwnerInTargetAssignment;
+
+    /**
+     * Handles all operations related to partition movements during a reassignment for balancing the target assignment.
+     */
+    private PartitionMovements partitionMovements;
+
+    /**
+     * The new assignment that will be returned.
+     */
+    private final Map<String, MemberAssignment> targetAssignment;
+
+    public GeneralUniformAssignmentBuilder(AssignmentSpec assignmentSpec, SubscribedTopicDescriber subscribedTopicDescriber) {
+        this.members = assignmentSpec.members();
+        this.subscribedTopicDescriber = subscribedTopicDescriber;
+        this.subscriptionIds = assignmentSpec.members().values().stream()
+            .flatMap(memberSpec -> memberSpec.subscribedTopicIds().stream())
+            .peek(topicId -> {
+                int partitionCount = subscribedTopicDescriber.numPartitions(topicId);
+                if (partitionCount == -1) {
+                    throw new PartitionAssignorException(
+                        "Members are subscribed to topic " + topicId + " which doesn't exist in the topic metadata."
+                    );
+                }
+            })
+            .collect(Collectors.toSet());
+        this.rackInfo = new RackInfo(assignmentSpec, subscribedTopicDescriber, subscriptionIds);
+        this.membersPerTopic = new HashMap<>();
+        members.forEach((memberId, memberMetadata) -> {
+            Collection<Uuid> topics = memberMetadata.subscribedTopicIds();
+            topics.forEach(topicId ->
+                membersPerTopic.computeIfAbsent(topicId, k -> new ArrayList<>()).add(memberId)
+            );
+        });
+        this.unassignedPartitions = new HashSet<>(allTopicIdPartitions(subscriptionIds, subscribedTopicDescriber));
+        this.assignedStickyPartitions = new HashSet<>();
+        this.assignmentManager = new AssignmentManager();
+        this.sortedMembersByAssignmentSize = assignmentManager.getSortedMembersByAssignmentSize(members.keySet());
+        this.currentPartitionOwners = new HashMap<>();
+        this.partitionOwnerInTargetAssignment = new HashMap<>();
+        this.targetAssignment = new HashMap<>();
+    }
+
     @Override
     protected GroupAssignment buildAssignment() {
-        return null;
+        if (subscriptionIds.isEmpty()) {
+            LOG.info("The subscription list is empty, returning an empty assignment");
+            return new GroupAssignment(Collections.emptyMap());
+        }
+
+        members.keySet().forEach(memberId -> targetAssignment.put(memberId, new MemberAssignment(new HashMap<>())));
```

Comment:
nit: Could you give me a suggestion for the formatting of this?

Reply:
i think what you have works. it's simple and it fits in a single line.

an alternative would be
```
        members.keySet().forEach(memberId -> 
            targetAssignment.put(memberId, new MemberAssignment(new HashMap<>()))
        );
```